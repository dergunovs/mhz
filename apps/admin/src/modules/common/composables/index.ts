import { useRoute } from 'vue-router';
import { useQueryClient } from '@tanstack/vue-query';

const queryClient = useQueryClient(); // стоит проверить, вызвав invalidate

export function useInvalidate(queryKey: string) {
  const route = useRoute(); 
  function invalidate() {
    if (route.query.create === '1') {
      queryClient.invalidateQueries({ queryKey: [queryKey], exact: true });
    }
  }

  return {
    invalidate,
  };
}
