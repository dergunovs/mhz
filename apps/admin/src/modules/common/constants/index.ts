export const URL_MAIN = '/main';
export const URL_ERROR = '/404';
