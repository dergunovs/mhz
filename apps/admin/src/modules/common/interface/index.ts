import { ICategory, IManager, IManufacturer, IProduct } from 'mhz-types';

export interface ISearchResult {
  products: IProduct[];
  categories: ICategory[];
  manufacturers: IManufacturer[];
  managers: IManager[];
}
