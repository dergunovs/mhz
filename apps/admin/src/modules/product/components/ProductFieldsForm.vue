<template>
  <div :class="$style.container">
    <h3>Specs</h3>

    <div v-for="field in formData" :key="field._id" data-test="product-fields-form-field">
      <UiCheckbox
        v-if="field.fieldType === 'boolean'"
        v-model="field.fieldValue"
        :label="field.title"
        data-test="product-fields-form-boolean"
      />

      <UiField v-else :label="field.fieldUnits ? `${field.title}, ${field.fieldUnits}` : field.title">
        <UiInput v-model="field.fieldValue" data-test="product-fields-form-value" />
      </UiField>
    </div>
  </div>
</template>

<script setup lang="ts">
import { onMounted, ref, watch } from 'vue';

import { UiCheckbox, UiField, UiInput } from 'mhz-ui';
import { ICategoryField } from 'mhz-contracts';
import { deleteTempId } from 'mhz-helpers';

interface IProps {
  fields: ICategoryField[];
  updates: number;
}

const props = defineProps<IProps>();
const emit = defineEmits<{ update: [formData: ICategoryField[]] }>();

const formData = ref<ICategoryField[]>([]);

watch(
  () => props.updates,
  () => {
    getFields();
  }
);

watch(
  () => formData.value,
  () => {
    emit('update', formData.value);
  },
  { deep: true }
);

function getFields() {
  formData.value = deleteTempId(props.fields, true);
}

onMounted(() => {
  getFields();
});
</script>

<style module lang="scss">
.container {
  display: flex;
  flex-direction: column;
  gap: 16px;
  width: 100%;
}
</style>
