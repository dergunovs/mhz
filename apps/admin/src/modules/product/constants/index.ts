export const URL_PRODUCT = '/products';
export const URL_PRODUCT_CREATE = `${URL_PRODUCT}/create`;
export const URL_PRODUCT_EDIT = `${URL_PRODUCT}/edit`;
