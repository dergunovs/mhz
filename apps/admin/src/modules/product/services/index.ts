import { Ref, ComputedRef } from 'vue';

import { useQuery, useMutation, IPageQuery } from 'mhz-helpers';
import { API_PRODUCT } from 'mhz-contracts';

import {
  getProductsApi,
  getProductApi,
  postProductApi,
  updateProductApi,
  deleteProductApi,
} from '@/product/services/api';

export function getProducts(query: Ref<IPageQuery | number>) {
  return useQuery({ queryKey: [API_PRODUCT, query], queryFn: () => getProductsApi(query) });
}

export function getProduct(id?: ComputedRef<string | string[]>) {
  return useQuery({ queryKey: [API_PRODUCT, id], queryFn: () => getProductApi(id) });
}

export function postProduct(options: object) {
  return useMutation({ mutationKey: [API_PRODUCT], mutationFn: postProductApi, ...options });
}

export function updateProduct(options: object) {
  return useMutation({ mutationKey: [API_PRODUCT], mutationFn: updateProductApi, ...options });
}

export function deleteProduct(options: object) {
  return useMutation({ mutationKey: [API_PRODUCT], mutationFn: deleteProductApi, ...options });
}
