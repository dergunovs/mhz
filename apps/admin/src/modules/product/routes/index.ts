import { URL_PRODUCT } from '@/product/constants';

export const productRoutes = [
  {
    path: URL_PRODUCT,
    name: 'ProductList',
    component: () => import('@/product/pages/ProductListPage.vue'),
  },
  {
    path: `${URL_PRODUCT}/:id`,
    name: 'Product',
    component: () => import('@/product/pages/ProductPage.vue'),
  },
];
