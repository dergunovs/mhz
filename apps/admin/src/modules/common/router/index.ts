import { createRouter, createWebHistory } from 'vue-router';

import { routes } from '@/common/router/routes';
import { getCookieToken } from '@/auth/composables';
import { URL_LOGIN } from '@/auth/constants';

const router = createRouter({
  history: createWebHistory(),
  routes,
  scrollBehavior(to, from, savedPosition) {
    return savedPosition ? savedPosition : { top: 0 };
  },
});

router.beforeEach((to, from, next) => {
  if (to.path !== URL_LOGIN && !getCookieToken()) {
    next({ path: `${URL_LOGIN}?logout=1` });
  } else {
    next();
  }
});

export default router;
