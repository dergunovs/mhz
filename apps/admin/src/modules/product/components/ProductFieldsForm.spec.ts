import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import * as helpers from 'mhz-helpers';
import { dataTest } from 'mhz-helpers';

import ProductFieldsForm from './ProductFieldsForm.vue';

import { wrapperFactory } from '@/common/test';
import { PRODUCTS } from '@/product/fixtures';

const spyDeleteTempId = vi.spyOn(helpers, 'deleteTempId');

const FIELDS = PRODUCTS.data[0].fields;

const productFieldsFormField = dataTest('product-fields-form-field');
const productFieldsFormBoolean = dataTest('product-fields-form-boolean');
const productFieldsFormValue = dataTest('product-fields-form-value');

let wrapper: VueWrapper;

beforeEach(() => {
  wrapper = wrapperFactory(ProductFieldsForm, {
    props: {
      fields: FIELDS,
      updates: 0,
    },
  });
});

enableAutoUnmount(afterEach);

describe('ProductFieldsForm', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(ProductFieldsForm)).toBeTruthy();
  });

  it('shows fields with specific components', async () => {
    expect(wrapper.findAll(productFieldsFormField).length).toEqual(FIELDS?.length);

    const booleanFieds = FIELDS?.filter((field) => field.fieldType === 'boolean').length;

    expect(wrapper.findAll(productFieldsFormBoolean).length).toEqual(booleanFieds);

    const valueFields = FIELDS?.filter((field) => field.fieldType === 'number' || field.fieldType === 'string').length;

    expect(wrapper.findAll(productFieldsFormValue).length).toEqual(valueFields);
  });

  it('emits updated form data by values update', async () => {
    expect(wrapper.emitted()['update']).toHaveLength(1);
    expect(wrapper.emitted()['update'][0]).toEqual([FIELDS]);

    await wrapper.findComponent(productFieldsFormBoolean).setValue(true);

    expect(wrapper.emitted()['update']).toHaveLength(2);
  });

  it('deletes ids of fields', async () => {
    expect(spyDeleteTempId).toBeCalledTimes(1);
    expect(spyDeleteTempId).toBeCalledWith(FIELDS, true);

    await wrapper.setProps({ updates: 1 });

    expect(spyDeleteTempId).toBeCalledTimes(2);
  });
});
