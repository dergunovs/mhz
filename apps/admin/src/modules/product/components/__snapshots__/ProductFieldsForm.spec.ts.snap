// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductFieldsForm > matches snapshot 1`] = `
"<div class="_container_11d7eb">
  <h3>Specs</h3>
  <div data-test="product-fields-form-field"><input type="checkbox" modelvalue="false" label="Onboard graphics" data-test="product-fields-form-boolean"></div>
  <div data-test="product-fields-form-field">
    <fieldset label="Size, Gb"><input type="text" modelvalue="32" data-test="product-fields-form-value"></fieldset>
  </div>
  <div data-test="product-fields-form-field">
    <fieldset label="Modules, pcs"><input type="text" modelvalue="2" data-test="product-fields-form-value"></fieldset>
  </div>
  <div data-test="product-fields-form-field">
    <fieldset label="Clock, MHz"><input type="text" modelvalue="6000" data-test="product-fields-form-value"></fieldset>
  </div>
  <div data-test="product-fields-form-field">
    <fieldset label="Type"><input type="text" modelvalue="DDR5" data-test="product-fields-form-value"></fieldset>
  </div>
</div>"
`;
